### ISSUE TITLE
BMI Code Refactoring Blueprints

### ISSUE THREAD

**Issue by apoorvemohan** [CCI-MOC/ims#96]
@chemistry-sourabh @ravisantoshgudimetla Are there any refactoring blueprints? If not, we should first create blueprint.

We would like to review them before we put any further effort in writing code (abstract classes etc.) for this.

**Comment by chemistry-sourabh** [comment]
First what do you want in the blueprint be as descriptive as possible ?

**Comment by chemistry-sourabh** [comment]
We'll probably need to meet again or do a hangout session atleast

**Comment by apoorvemohan** [comment]
By blueprint i meant a design document.

**Comment by ravisantoshgudimetla** [comment]
I think I commented in other issue as well. I guess, the easiest thing to do is define at high level what we want to do and create a  template. Everytime, a new design proposal is to made, we need to fill in, have that reviewed and then proceed with coding. @apoorvemohan if you can make an initial pass, it would be great. I can lend a hand if needed and I believe @chemistry-sourabh is willing to do that as well.

**Comment by chemistry-sourabh** [comment]
I can come to Boston next weekend or we can have a hangout session sometime evening this evening if its urgent

**Comment by apoorvemohan** [comment]
@ravisantoshgudimetla The interfaces you are designing, are they based on any previous discussion? If yes, could you share the document that summarises the design discussions? 

If not, we should first come up with a proper design document before we start writing any code.  

The goal here is to have the interns work on code refactoring during the summer (guided by @naved001). For that, we need to prepare a good design document by end of April (this will include the flow diagram, DB schema, the list of API's exposed at the different interfaces and designing the interfaces).

**Comment by ravisantoshgudimetla** [comment]
@apoorvemohan We never had anything formal AFAIK but I guess this was something that has been long discussed probably since last summer. 

>If not, we should first come up with a proper design document before we start writing any code.
+1. I just don't understand the format and the way you want it to be. So, make a initial pass on the format and let the discussion get started. We can refine over it in coming weeks.

**Comment by ravisantoshgudimetla** [comment]
Accidentally closed. Reopened it now.

**Comment by apoorvemohan** [comment]
Ok.

**Comment by knikolla** [comment]
Please propose the document designs into the code repository so people can do code review on them and comment line by line. Similar to the openstack specs process.

**Comment by ravisantoshgudimetla** [comment]
@knikolla Thanks for the spec documentation. I guess it may be overkill as we don't have too many components. At this point, given that we have too many external contributors and everyone is not familiar with using something like that(including me) I suggest to have a stripped down version of that template, perhaps that we can use internally. Something like this:

- Create a md file with some fields taken from the template you provided.
- Create a PR for the proposal against either ims repo(?) or proposals repo(?- a new repo for proposals)
- Have people review it and provide feedback
- Merge it in the repo which means its locked.
- Create issues based on the merged proposal.

I am completely open to other suggestions too if you or others have any other feedback.

**Comment by chemistry-sourabh** [comment]
I agree with @ravisantoshgudimetla, but instead of creating an md file in the repo why not in the wiki ? and raise new proposal PRs to the wiki ?

I advice @apoorvemohan to assign this to someone (Not me :P) and let that person make simplified version of @knikolla's link and raise a PR for the Wiki.

**Comment by knikolla** [comment]
@chemistry-sourabh docs are part of your main repo, not the wiki repo. also the wiki repo is too github specific. i don't have strong opinions on this anyway, as long as there is an open formal review and approval process.
